alphacsc/encoder.py

```diff
@@ -0,0 +1,194 @@
+from .loss_and_gradient import compute_X_and_objective_multi
+from .update_z_multi import update_z_multi
+from .utils import check_dimension, lil
+
+# XXX check consistency / proper use!
+def get_z_encoder_for(solver, z_kwargs, X, D_hat, n_atoms, atom_support, algorithm, reg, loss, loss_params, uv_constraint, feasible_evaluation, n_jobs):
+    """
+    Returns a z encoder for the required solver.
+    Allowed solvers are ['l-bfgs', 'lgcd', dicodile']
+
+    Parameters
+    ----------
+    algorithm : 'batch' | 'greedy' | 'online' | 'stochastic'
+        Dictionary learning algorithm.
+
+    Returns
+    -------
+    enc - a ZEncoder instance
+
+    Example usage
+    -------------
+    with get_encoder_for('dicodile') as enc:
+        ...
+    """
+    if solver == 'dicodile':
+        assert loss == 'l2'
+        return DicodileEncoder(X, n_workers=n_jobs) 
+    elif solver in ['l-bfgs', 'lgcd']:
+        return AlphaCSCEncoder(solver, z_kwargs, X, D_hat, n_atoms, atom_support, algorithm, reg, loss, loss_params, uv_constraint, feasible_evaluation, n_jobs)
+    else:
+        raise ValueError(f'unrecognized solver type: {solver}.')
+
+
+class BaseZEncoder:
+    def compute_z(self):
+        """
+        """
+        raise NotImplementedError()
+    
+    def compute_z_partial(self, i0):
+        """
+        (Online learning)
+        """
+        raise NotImplementedError()
+
+    def get_cost(self):
+        """
+        Returns
+        -------
+        cost: float
+        """
+        raise NotImplementedError()
+
+    def get_sufficient_statistics(self):
+        """
+        Returns
+        -------
+        ztz, ztX
+        """
+        raise NotImplementedError()
+
+    def get_z_hat(self):
+        raise NotImplementedError()
+    
+    def get_z_hat_partial(self, i0):
+        """
+        (Online learning)
+        """
+        raise NotImplementedError()
+
+    def set_D(self, d):
+        """
+        Update the dictionary
+        """
+        raise NotImplementedError()
+    
+    def set_reg(self, reg):
+        """
+        Update the regularization parameter
+        """
+        raise NotImplementedError()
+    
+    def add_one_atom_in_z(self):
```

Comment:
Maybe just use `add_one_atom(self, d_k)` that should add an atom in `D_hat` and add a 0 line in `z_hat`